optuna/samplers/_tpe/sampler.py

```diff
@@ -204,34 +283,64 @@ def infer_relative_search_space(
+            for sub_space in self._search_space_group.search_spaces:
+                for name, distribution in sub_space.items():
+                    if not isinstance(distribution, _DISTRIBUTION_CLASSES):
+                        self._log_independent_sampling(n_complete_trials, trial, name)
+                        continue
+                    search_space[name] = distribution
+            return search_space
+
         for name, distribution in self._search_space.calculate(study).items():
             if not isinstance(distribution, _DISTRIBUTION_CLASSES):
-                if self._warn_independent_sampling:
-                    complete_trials = study.get_trials(deepcopy=False)
-                    if len(complete_trials) >= self._n_startup_trials:
-                        self._log_independent_sampling(trial, name)
+                self._log_independent_sampling(n_complete_trials, trial, name)
                 continue
             search_space[name] = distribution
 
         return search_space
 
-    def _log_independent_sampling(self, trial: FrozenTrial, param_name: str) -> None:
-        _logger.warning(
-            "The parameter '{}' in trial#{} is sampled independently "
-            "instead of being sampled by multivariate TPE sampler. "
-            "(optimization performance may be degraded). "
-            "You can suppress this warning by setting `warn_independent_sampling` "
-            "to `False` in the constructor of `TPESampler`, "
-            "if this independent sampling is intended behavior.".format(param_name, trial.number)
-        )
+    def _log_independent_sampling(
+        self, n_complete_trials: int, trial: FrozenTrial, param_name: str
+    ) -> None:
+        if self._warn_independent_sampling:
+            if n_complete_trials >= self._n_startup_trials:
+                _logger.warning(
+                    f"The parameter '{param_name}' in trial#{trial.number} is sampled "
+                    "independently instead of being sampled by multivariate TPE sampler. "
+                    "(optimization performance may be degraded). "
+                    "You can suppress this warning by setting `warn_independent_sampling` "
+                    "to `False` in the constructor of `TPESampler`, "
+                    "if this independent sampling is intended behavior."
+                )
 
     def sample_relative(
         self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
     ) -> Dict[str, Any]:
 
         self._raise_error_if_multi_objective(study)
 
+        if self._group:
+            assert self._search_space_group is not None
+            params = {}
+            for sub_space in self._search_space_group.search_spaces:
+                for name, distribution in sub_space.items():
+                    if not isinstance(distribution, _DISTRIBUTION_CLASSES):
+                        del sub_space[name]
```

Comment:
You cannot change the size of the container while iterating over it. Let's create a new/copy dictionary of `sub_space`.